opal/corelib/array.rb

```diff
@@ -1357,16 +1368,27 @@ def keep_if(&block)
     return enum_for(:keep_if) { size } unless block_given?
 
     %x{
-      for (var i = 0, length = self.length, value; i < length; i++) {
-        value = block(self[i]);
+      var value, raised = null, updated = new Array(self.length);
 
-        if (value === false || value === nil) {
-          self.splice(i, 1);
+      for (var i = 0, i2 = 0, length = self.length; i < length; i++) {
+        try {
+          value = Opal.yield1(block, self[i])
+        } catch(error) {
```

Comment:
Also I guess it may catch and re-throw `Opal.breaker` and `Opal.returner` and thus, potentially exit the outer method

Reply:
From what I saw they're already managed by the compiler.

Reply:
compiler manages only what is written in Ruby (but catch/re-throw here is in JS). maybe I'm wrong, but does it handle `break` and `return` inside `keep_if`? 👍 if so, but I'm afraid it swallows the initial `throw Opal.returner` and re-raises it once the iteration is done (MRI most probably does a hard `return` that aborts the loop)

Also, I'm not sure if it's actually that important. I've never seen such code 😄

Reply:
just out of curiosity, here's what I got:

```
opal:elia/fix-delete-if-keep-if ⤑ bin/opal -e'def a;4.times{|n| [1,2,3].delete_if { |i| p i; return }; p n => n};end; a'          ~/C/opal
1
opal:elia/fix-delete-if-keep-if ⤑ ruby -e'def a;4.times{|n| [1,2,3].delete_if { |i| p i; return }; p n => n};end; a'              ~/C/opal
1
opal:elia/fix-delete-if-keep-if ⤑                                                                                                 ~/C/opal
```